### ISSUE TITLE
With arc/orc, created thread crashes when the `Thread[T]` variable goes out of scope

### ISSUE THREAD

**Issue by ChengCat** [nim-lang/Nim#18774]
When compiled with `nim c --threads:on --gc:arc a.nim`, this program crashes:

```
import os

proc entry(env: string) =
   echo env

proc f1(env: string) =
   var thrd: Thread[string]
   thrd.createThread(entry, env)

f1("abc")
sleep(1000)
```

The old GC works fine for the same program. A slightly modified program also works fine with arc/orc:

```
import os

proc entry(env: string) =
   echo env

var thrd: Thread[string]
thrd.createThread(entry, "abc")
sleep(1000)
```

This is a somewhat surprising behavior to me. It's probably due to some memory freed when `thrd` goes out of scope. I don't need to keep `thrd` around, because the thread is meant to run forever.

Tested with the latest devel commit: `3469f3a39320cf35d02d11b78c7db0be79e992f8`

**Comment by Araq** [comment]
Threads in Nim could never reliably allocated on the stack. You always need to use your workaround. Or use `spawn` as `createThread` is a low level construct. And yeah, a new threading API would be nice.

**Comment by ChengCat** [comment]
Please make this clear in the documents then. This was very confusing to me. I honestly can't think of any other language would either crash or subtly corrupt memory for doing this. For example, in C11 `thrd_t` is defined as "a complete object type that holds an identifier for a thread" in the standard.

I'd expect this issue very simple to fix in Nim: you can defer the time to free whatever memory `Thread[T]` holds, to the time when the thread is joined. If a thread is never joined, the memory doesn't have to be freed.

I didn't use `spawn` because the documentation says it's unstable, and I can't wait on `FlowVar[T]` and `Channel[T]` simultaneously. I am making a multi-threaded program with relatively complex synchronizations, and I've found `Thread` and `Channel` together is best approach for me.

**Comment by Araq** [comment]
> I'd expect this issue very simple to fix in Nim

Feel free to create a PR.

**Comment by ChengCat** [comment]
> Feel free to create a PR.

I'm happy to contribute! For the time being, I want to finish my own project as soon as possible. I will take a closer look soon.

EDIT: Sorry, I don't use Nim any more. I can help improving the language, but I personally can't handle having to debate with Araq almost trivial problems like this, each time after I get bitten by the language. My project is also blocked by another ~5 years old bug in Nim standard library, and that bug is difficult to properly fix.
